### PR TITLE
Adopt #13446 server favorite button

### LINKED ISSUE #12647: Add favorite button
## Problem

I can't add server to Favorites list.
When i move to new MT version, i have to quick visit each server to have it in favorites :/

## Solutions

Add button "Add to favorites" (image_button/whatever, same as Remove favorite)

## Additional context

Related #4028

### PR DESCRIPTION
**Pull request by siliconsniffer**
This PR adds a button for adding a server to the favorites.

- Goal of the PR
Add a button to mark server as favorite without needing to join.

- Does it resolve any reported issue?
Resolves #12647

- Does this relate to a goal in [the roadmap](https://github.com/minetest/minetest/blob/master/doc/direction.md)?
2.3 I'd say.

- If not a bug fix, why is this PR needed? What usecases does it solve?
One won't need to connect to a server to mark it as favorite.

## Preview
image
image

## How to test

Mark & remove as favorite, try while searching and check nothing else broke.
Check that the selection is updating.

### CODE REVIEW

File: builtin/mainmenu/tab_online.lua [hunk 1]
```diff
@@ -331,6 +341,14 @@ local function main_button_handler(tabview, fields, name, tabdata)
 		end
 	end
 
+	if fields.btn_add_favorite then
+		local server = find_selected_server()
+
+		serverlistmgr.add_favorite(server)
+		set_selected_server(server)
```
Comment (appgurueu) [inline builtin/mainmenu/tab_online.lua:348]: This shouldn't be necessary. Is it?

File: builtin/mainmenu/tab_online.lua [hunk 2]
```diff
@@ -169,6 +169,16 @@ local function get_formspec(tabview, name, tabdata)
 			retval = retval .. "style[btn_delete_favorite;padding=6]"
 			retval = retval .. "image_button[" .. (can_view_clients_list and "4.5" or "5") .. ",1.3;0.5,0.5;" ..
 				core.formspec_escape(defaulttexturedir .. "server_favorite_delete.png") .. ";btn_delete_favorite;]"
+		else
+			retval = retval .. "tooltip[btn_add_favorite;" .. fgettext("Add favorite") .. "]"
+			retval = retval .. "style[btn_add_favorite;padding=6]"
+			if not can_view_clients_list then
```
Comment (appgurueu) [inline builtin/mainmenu/tab_online.lua:175]: As ywang has suggested, you can use a ternary to vary just the X coordinate below.

File: builtin/mainmenu/tab_online.lua (hunk 1 above)
Comment (siliconsniffer) [reply builtin/mainmenu/tab_online.lua:348]: I think it is, removing it won't update the selection after the server is marked as favorite.

Comment (appgurueu) [reply builtin/mainmenu/tab_online.lua:348]: > I think it is, removing it won't update the selection after the server is marked as favorite.

But you can only mark an already selected server as favorite?

In other words: If `find_selected_server` returns something that is not `nil`, `server` is already selected and `set_selected_server(server)` should be unnecessary.

Comment (appgurueu) [reply builtin/mainmenu/tab_online.lua]: Update after discussion on Discord: Turns out it is necessary for the side effect of updating `gamedata.fav`. Frankly we should just remove `gamedata.fav` eventually (it can be derived from the selected server) but for now this PR should simply set `gamedata.fav = true` here instead.

File: builtin/mainmenu/tab_online.lua [hunk 3]
```diff
@@ -331,6 +336,14 @@ local function main_button_handler(tabview, fields, name, tabdata)
 		end
 	end
 
+	if fields.btn_add_favorite then
+		local server = find_selected_server()
+
+		serverlistmgr.add_favorite(server)
```
Comment (appgurueu) [inline builtin/mainmenu/tab_online.lua:342]:
```suggestion
		serverlistmgr.add_favorite(find_selected_server())
```
(and then get rid of the `local server` right above)

**Review by appgurueu — APPROVED** [review]
Works, code is fine

File: builtin/mainmenu/tab_online.lua [hunk 4]
```diff
@@ -56,18 +56,23 @@ local function get_sorted_servers()
...
-	gamedata.serverdescription = server.description
 
-	gamedata.fav = false
-	for _, fav in ipairs(serverlistmgr.get_favorites()) do
-		if address == fav.address and port == fav.port then
-			gamedata.fav = true
-			break
-		end
-	end
+	is_selected_fav()
```
Comment (appgurueu) [inline builtin/mainmenu/tab_online.lua:75]:
```suggestion
```

After this refactor, nothing needs to be done here: `is_selected_fav` is just determined on demand, not eagerly updated here.

File: builtin/mainmenu/tab_online.lua [hunk 5]
```diff
@@ -56,18 +56,23 @@ local function get_sorted_servers()
...
+
+local function is_selected_fav(server)
+	local address = core.settings:get("address")
+	local port = tonumber(core.settings:get("remote_port"))
+
+	for _, fav in ipairs(serverlistmgr.get_favorites()) do
+			if address == fav.address and port == fav.port then
+				return true
+			end
+		end
```
Comment (appgurueu) [inline builtin/mainmenu/tab_online.lua:68]: 1. Indentation is wrong here (need to dedent the preceding 4 lines by 1)
2. I would add a `return false` for minor style reasons (conceptually this returns a bool: whether the selected server is a favorite; nil technically has the correct falsiness but it doesn't sit right with me considering our code style)

File: builtin/mainmenu/tab_online.lua [hunk 6]
```diff
@@ -56,18 +56,23 @@ local function get_sorted_servers()
...
+end
+
 local function set_selected_server(server)
 	local address = server.address
 	local port    = server.port
-	gamedata.serverdescription = server.description
```
Comment (appgurueu) [inline builtin/mainmenu/tab_online.lua:62]: Unrelated change which might cause minor regressions, please either:

* Revert, *or*
* Refactor `gamedata.serverdescription` away in a similar manner as `gamedata.fav`: There only is one usage, and that usage should probably just use `server.description or ""` where `local server = find_selected_server()`.

Comment (siliconsniffer) [reply builtin/mainmenu/tab_online.lua:62]: Will revert :+1:

File: builtin/mainmenu/tab_online.lua [hunk 7]
```diff
@@ -56,18 +56,23 @@ local function get_sorted_servers()
 end
 
 -- Persists the selected server in the "address" and "remote_port" settings
```
Comment (appgurueu) [inline builtin/mainmenu/tab_online.lua:58]: This comment belongs to `set_selected_server`. Either update it (so it applies to multiple of the functions below) or move it there.

**Review by appgurueu — APPROVED** [review]
~~Only a very minor comment left, otherwise~~ LGTM!

**Comment by siliconsniffer** [comment]
Maybe one last thing to consider is how to handle the removal of servers.
In my opinion no. 1 is more annoying. Obviously option no. 2 could also go to the server above the selected. 

https://github.com/user-attachments/assets/e430e8fa-1976-48d8-af13-39c99d9035f7


https://github.com/user-attachments/assets/fd9bd837-1fd2-4fa8-a144-f0fac78315b7

**Comment by appgurueu** [comment]
> Maybe one last thing to consider is how to handle the removal of servers.

Either is fine by me. No need to overthink this.

(Also, isn't this technically out of scope for this PR, which ought to be just adding a way to favorite servers?)

**Comment by siliconsniffer** [comment]
> (Also, isn't this technically out of scope for this PR, which ought to be just adding a way to favorite servers?)

Probably, yes. Let's merge it like it is and wait for someone to complain.

**Comment by siliconsniffer** [comment]
Misleading commit message, the button is now positioned on the right end of the container instead of being left of the clients  button.

**Comment by appgurueu** [comment]
> Misleading commit message, the button is now positioned on the right end of the container instead of being left of the clients button.

In such a case, just do `git commit --amend`, fix the commit message, then `git push -f` (force push).

But it doesn't really matter since we're gonna squash this anyways.

**Review by SmallJoker — APPROVED** [review]
Works.
